### PR TITLE
feat: configurable chat history scrollback

### LINKED ISSUE #422: feat: configurable chat history scrollback (avoid losing older messages)
## Problem
In long-running threads, older chat history becomes inaccessible in the UI: as a conversation grows, the beginning of the thread disappears from the scrollback.

This makes it hard to recover context in longer chats and power-user sessions.

## Current behavior
- The UI retains only a small window of conversation items per thread.
- As more items arrive (streaming deltas, tool output, etc.), earlier items are dropped locally.

## Root cause (code)
The frontend unconditionally limits thread items via `prepareThreadItems()`:
- `src/utils/threadItems.ts` has `MAX_ITEMS_PER_THREAD = 200`.
- `prepareThreadItems()` slices to the last 200 items (`normalized.slice(-MAX_ITEMS_PER_THREAD)`).
- Reducers call `prepareThreadItems(...)` during normal thread updates, so the cap applies continuously.

## Proposal
Add an **app-wide** setting to configure thread history scrollback retention:
- Default stays **200** items (no behavior change for most users).
- Allow power users to increase to higher values (e.g. 500/1000/2000/5000).
- Optional **Unlimited** mode (stores all items returned by `thread/resume` + live items).

## UX
Settings: `Settings -> Display & Sound` (or similar)
- Toggle: `Unlimited chat history`
- Number input: `Max items per thread` (disabled when Unlimited is on)
- Presets: 200 / 500 / 1000 / 2000 / 5000
- Help text warning about performance/memory costs for large values.

Note: existing thread context menu already has **“Sync from server”**; after increasing scrollback, users can sync to pull additional older history.

## Performance considerations
Increasing retention can increase memory/DOM/render cost. To mitigate:
- Virtualize the message list (`@tanstack/react-virtual` is already in the repo and used elsewhere).
- Virtualize at the grouped-item level (tool groups + items), with ResizeObserver measurement like `GitDiffViewer`.

## Acceptance criteria
- Default remains last-200 behavior.
- Setting persists and applies to new and existing threads.
- Higher limits retain more items; Unlimited retains all items received.
- No major regressions in scroll/auto-scroll behavior.
- Tests cover retention logic (default/custom/unlimited) and settings normalization.

### PR DESCRIPTION
**Pull request by amanthanvi**
Fixes #422

## What
- Adds an **app-wide** setting for chat history scrollback (max items retained per thread).
- Adds a Settings UI (unlimited toggle, presets, custom number input) with guidance about performance + re-syncing older messages.
- Plumbs the setting into thread item retention logic (null = unlimited).
- Virtualizes the conversation list to reduce render cost for large histories (uses existing `@tanstack/react-virtual`).

## Notes
- On startup, trimming is deferred until persisted settings hydrate (scrollback is treated as unlimited briefly), preventing older items from being dropped before the configured limit is known.
- Increasing the scrollback limit does not automatically re-fetch older history that was previously trimmed; use **"Sync from server"** on a thread to restore older messages.
- "Unlimited" may increase memory usage; virtualization mitigates UI rendering cost but not memory growth.

## How to validate
- Open Settings → Display & Sound → Chat and adjust scrollback (toggle Unlimited, pick presets, or enter a custom value).
- Confirm longer threads retain more items locally and UI remains responsive with large histories.

## Tests
- `npm run lint`
- `npm run typecheck`
- `npm run test`
- `cd src-tauri && cargo test`

### CODE REVIEW

**Comment by amanthanvi** [comment]
@codex review

File: src/features/settings/components/sections/SettingsDisplaySection.tsx [hunk 1]
```diff
@@ -171,6 +272,88 @@ export function SettingsDisplaySection({
...
+      </div>
+      <div className="settings-field">
+        <label className="settings-field-label" htmlFor="chat-scrollback-preset">
+          Scrollback preset
+        </label>
+        <select
+          id="chat-scrollback-preset"
+          className="settings-select"
+          value={scrollbackPresetValue}
+          onChange={(event) => selectScrollbackPreset(event.target.value)}
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/features/settings/components/sections/SettingsDisplaySection.tsx:306]: Fixed in e8bf542: disable the preset dropdown while Unlimited is enabled, and added a guard so preset changes cannot flip Unlimited off. Added a regression test for the disabled controls.

File: src/features/settings/components/sections/SettingsDisplaySection.tsx [hunk 2]
```diff
@@ -171,6 +272,88 @@ export function SettingsDisplaySection({
...
+            onClick={() => {
+              setScrollbackDraft(String(CHAT_SCROLLBACK_DEFAULT));
+              void onUpdateAppSettings({
+                ...appSettings,
+                chatHistoryScrollbackItems: CHAT_SCROLLBACK_DEFAULT,
+              });
+            }}
+          >
+            Reset
+          </button>
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/features/settings/components/sections/SettingsDisplaySection.tsx:355]: Fixed in e8bf542: disable the scrollback Reset button while Unlimited is enabled (consistent with the input). Covered by the same regression test.

File: src/features/messages/components/Messages.tsx [hunk 3]
```diff
@@ -442,78 +440,221 @@ export const Messages = memo(function Messages({
...
+      }
+      rowNodesByKeyRef.current.set(key, node);
+      rowVirtualizer.measureElement(node);
+      if (rowResizeObserversRef.current.has(node)) {
+        return;
+      }
+      const observer = new ResizeObserver(() => {
+        rowVirtualizer.measureElement(node);
+        requestAutoScroll();
+      });
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/features/messages/components/Messages.tsx:626]: Kept the standard new ResizeObserver(callback) usage. The vitest ResizeObserverMock in src/test/vitest.setup.ts tolerates extra args at runtime, and CI/typecheck are green. I prefer avoiding the suggested shim unless we hit a real incompatibility.

File: src/features/messages/components/Messages.tsx [hunk 4]
```diff
@@ -442,78 +440,221 @@ export const Messages = memo(function Messages({
...
-          const groupBodyId = `tool-group-${group.id}`;
-          const ChevronIcon = isCollapsed ? ChevronDown : ChevronUp;
           return (
             <div
-              key={`tool-group-${group.id}`}
-              className={`tool-group ${isCollapsed ? "tool-group-collapsed" : ""}`}
+              key={entry.key}
+              className="messages-virtual-row"
+              data-index={virtualRow.index}
+              ref={setRowRef(entry.key)}
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/features/messages/components/Messages.tsx]: Addressed in 7b9fb13: cache ref callbacks per virtual row key to avoid ref detach/attach churn (and thus ResizeObserver churn) on rerenders. Added a regression test asserting ResizeObserver.observe is not called again on rerender.

File: src/features/threads/hooks/useThreads.ts [hunk 5]
```diff
@@ -55,11 +56,18 @@ export function useThreads({
...
   customPrompts = [],
   onMessageActivity,
   threadSortKey = "updated_at",
 }: UseThreadsOptions) {
   const [state, dispatch] = useReducer(threadReducer, initialState);
+  const maxItemsPerThread =
+    chatHistoryScrollbackItems === undefined ? 200 : chatHistoryScrollbackItems;
+
+  useEffect(() => {
+    dispatch({ type: "setMaxItemsPerThread", maxItemsPerThread });
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/features/threads/hooks/useThreads.ts:76]: Good catch — fixed in 21fc843.

- Seed `maxItemsPerThread` into the reducer initial state (useReducer initializer) so we don’t start at 200 and rely on a follow-up effect.
- In App boot, pass `chatHistoryScrollbackItems=null` while `appSettingsLoading` so we don’t trim before persisted settings hydrate (later dispatch can safely trim down, but we never lose items by trimming too early).
- Added an integration test covering “load >200 items before hydration, then apply 200 and verify trimming”.

**Comment by amanthanvi** [comment]
Follow-ups pushed to address review feedback:
- Disable preset + reset while Unlimited is enabled (e8bf542).
- Stabilize per-row ref callbacks to avoid ResizeObserver churn on rerender (7b9fb13).
- Fix settings hydration edge case so we don't trim history before persisted scrollback loads; reducer now seeds maxItemsPerThread up-front and trimming is deferred during settings load (21fc843, plus regression test).

CI is green (lint/typecheck/test + Rust + Tauri builds).

**Comment by amanthanvi** [comment]
@codex review

**Comment by Dimillian** [comment]
@codex review

**Comment by Dimillian** [comment]
@amanthanvi it introduces an issue, when loading an old thread, and trying to scrollup it'll jump back to the bottom making it almost impossible to scroll up

**Comment by Dimillian** [comment]
Overall lot of complexity by virtualizing the conversation view

**Comment by amanthanvi** [comment]
Got it. Hoping a nice way to add this can be figured out somewhere down the line. Appreciate you taking a look at it! @Dimillian

**Comment by amanthanvi** [comment]
Follow-up: opened #427 as a reduced-scope alternative (same configurable scrollback setting + retention logic, but no Messages virtualization) based on the scroll-jump + complexity feedback here.
